proxy/envoy/config.go

```diff
@@ -226,15 +290,64 @@ func buildListeners(instances []*model.ServiceInstance,
 		hosts := make(map[string]VirtualHost, 0)
 		for _, proto := range []model.Protocol{model.ProtocolHTTP, model.ProtocolHTTP2, model.ProtocolGRPC} {
 			for _, svc := range lst.services[proto] {
-				host := buildHost(svc, OutboundClusterPrefix+svc.String(), suffix)
+				// if there is a routing rule for the host, use that to generate the route block
+				routes := make([]Route, 0)
+				// TODO: do we match on partial hostnames (foo.svc, foo.svc.namespace, etc. ?)
+				ruleByDestination, prs := rulesMapTmp[svc.Hostname]
+				if prs {
+					for _, rule := range ruleByDestination {
+						httpRule := rule.GetHttp()
+						route := Route{}
+						if httpRule != nil {
+							match := httpRule.GetMatch()
+							if match != nil && match.GetUri() != nil {
+								// TODO Error check. Either path/prefix, but not both
+								// TODO Error check. No regex
+								route.Prefix = match.Uri.GetPrefix()
+								route.Path = match.Uri.GetExact()
+								if route.Prefix == "" && route.Path == "" {
+									route.Prefix = "/"
+								}
+							}
+							if httpRule.WeightedClusters != nil {
+								// Need to populate the weighted_clusters block with the
+								// cluster name of the upstream cluster and the weight.
+								// The user provides the destination service and tags and
+								// we have to convert this into the appropriate upstream cluster name
+								wcRule := httpRule.WeightedClusters
+								route.WeightedClusters = &WeightedCluster{}
+								route.WeightedClusters.Clusters = make([]WeightedClusterEntry, 0)
+								for _, wcRuleItem := range wcRule {
+									wcEntry := WeightedClusterEntry{
+										Name: OutboundClusterPrefix +
+											clusterIdentifierMap[clusterIdentifierString(wcRuleItem.DstCluster)],
+										Weight: int(wcRuleItem.Weight),
+									}
+									route.WeightedClusters.Clusters = append(route.WeightedClusters.Clusters, wcEntry)
+								}
+							} else {
+								route.Cluster = OutboundClusterPrefix + svc.String()
+							}
+						}
+						// no else. Because a L4 only based match for a HTTP service makes no sense for Envoy,
+						// as it does not offer the ability to filter HTTP traffic by src/dst IP.
+						// It is the mixer's job to police such traffic.
+						routes = append(routes, route)
+					}
+				} else {
+					routes = append(routes, Route{Prefix: "/", Cluster: OutboundClusterPrefix + svc.String()})
+				}
+				host := buildHost(svc, suffix)
+				host.Routes = routes
```

Comment:
use functional style: pass a parameter to `buildHost` unless it's absolutely necessary to mutate the object.